src/routers/StackRouter.js

```diff
@@ -100,63 +155,11 @@ export default (routeConfigs, stackConfig = {}) => {
     getStateForAction(action, state) {
       // Set up the initial state if needed
       if (!state) {
-        let route = {};
-        const childRouter = childRouters[action.routeName];
-        // This is a push-like action, and childRouter will be a router or null if we are responsible for this routeName
-        if (behavesLikePushAction(action) && childRouter !== undefined) {
-          let childState = {};
-          // The router is null for normal leaf routes
-          if (childRouter !== null) {
-            const childAction =
-              action.action ||
-              NavigationActions.init({ params: action.params });
-            childState = childRouter.getStateForAction(childAction);
-          }
-          return {
-            key: 'StackRouterRoot',
-            isTransitioning: false,
-            index: 0,
-            routes: [
-              {
-                params: action.params,
-                ...childState,
-                key: action.key || generateKey(),
-                routeName: action.routeName,
-              },
-            ],
-          };
-        }
-
-        if (initialChildRouter) {
-          route = initialChildRouter.getStateForAction(
-            NavigationActions.navigate({
-              routeName: initialRouteName,
-              params: initialRouteParams,
-            })
-          );
-        }
-        const params = (route.params ||
-          action.params ||
-          initialRouteParams) && {
-          ...(route.params || {}),
-          ...(action.params || {}),
-          ...(initialRouteParams || {}),
-        };
-        route = {
-          ...route,
-          ...(params ? { params } : {}),
-          routeName: initialRouteName,
-          key: action.key || generateKey(),
-        };
-        state = {
-          key: 'StackRouterRoot',
-          isTransitioning: false,
-          index: 0,
-          routes: [route],
-        };
+        return getInitialState(action);
```

Comment:
I like this. I wonder if we should make this file way more functiony, in the name of comprehension. Create functions for each action and attempt to get `getStateForAction` down to ~100 lines or something